### PR TITLE
Add OptionalObjectToStringConverterFactory to support Optional parameters in retrofit2 clients, fixes #326

### LINKED ISSUE #326: Add Java8-Optional functionality to retrofit2-clients

### PR DESCRIPTION
**Pull request by uschi2000**
…ters in retrofit2 clients, fixes 326

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/331)

### CODE REVIEW

File: retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/OptionalObjectToStringConverterFactory.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+        return null;
+    }
+
+    enum Java8OptionalStringConverter implements Converter<java.util.Optional<?>, String> {
+        INSTANCE;
+
+        @Override
+        public String convert(java.util.Optional<?> value) throws IOException {
+            return value.map(Object::toString).orElse("");
```
Comment (markelliot) [inline retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/OptionalObjectToStringConverterFactory.java:64]: don't we want to nest the mapping here? ie this should be non-empty optionals get converted using the converter of appropriate type for the inner type?

Comment (markelliot) [reply retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/OptionalObjectToStringConverterFactory.java:64]: (say by using the JacksonConverter?)

Comment (uschi2000) [reply retrofit2-clients/src/main/java/com/palantir/remoting1/retrofit2/OptionalObjectToStringConverterFactory.java:64]: Arguably, all query or path params should be mostly stringish thing anyway, no?

**Comment by uschi2000** [comment]
I'd also prefer not diverging from plain retrofit2 behavior too much, if
possible. out-of-the-box behavior is to serialize everything as #toString,
even if you apply the JacksonConverterFactory.

**Comment by markelliot** [comment]
What about return types?

**Comment by markelliot** [comment]
Suppose I have some API:

interface Blargh {
    Optional<AliasType> getThing();
}

For:

class AliasType {
    private final int val;

    @JsonCreator
    public AliasType(int val) {
        this.val = val;
    }

    @JsonValue
    public int get() {
        return val;
    }
}

Will this work?

**Comment by uschi2000** [comment]
I think so, let me try.

**Comment by uschi2000** [comment]
yep, see test.

**Pull request merged**